mlir/lib/Dialect/LLVMIR/IR/LLVMLinkerInterface.cpp

```diff
@@ -11,30 +11,435 @@
+
+  LLVMSymbolLinkerInterface(Dialect *dialect) : Base(dialect) {}
+
+  enum class LinkFrom { Dst, Src, Both };
+
+  ModuleLinkerInterface *getModuleLinkerInterface(Operation *op) const {
+    ModuleOp mod = op->getParentOfType<ModuleOp>();
+    if (!mod)
+      return nullptr;
+    return dyn_cast_or_null<ModuleLinkerInterface>(mod->getDialect());
+  }
+
+  bool canBeLinked(Operation *op) const override {
+    return isa<LLVM::GlobalOp>(op) || isa<LLVM::LLVMFuncOp>(op);
+  }
+
+  StringRef getSymbol(Operation *op) const override { return symbol(op); }
+
+  ConflictPair findConflict(Operation *src) const override {
+    assert(canBeLinked(src) && "expected linkable operation");
+
+    if (isLocalLinkage(getLinkage(src)))
+      return {nullptr, src};
+
+    // TODO: make lookup through module state
+    if (ModuleLinkerInterface *iface = getModuleLinkerInterface(src)) {
+      if (Operation *dst = lookup(src)) {
+        if (dst != src && !isLocalLinkage(getLinkage(dst))) {
+          return {dst, src};
+        }
+      }
+    }
+
+    return {nullptr, src};
+  }
+
+  bool isLinkNeeded(ConflictPair pair) const override {
+    assert(canBeLinked(pair.src) && "expected linkable operation");
+
+    LLVM::Linkage srcLinkage = getLinkage(pair.src);
+
+    // Always import variables with appending linkage.
+    if (isAppendingLinkage(srcLinkage)) {
+      return true;
+    }
+
+    if (shouldLinkOnlyNeeded()) {
+      // Don't import globals that are already defined
+      if (!pair.dst || !isDeclaration(pair.dst))
+        return false;
+    }
+    if (isDeclaration(pair.src))
+      return false;
+
+    bool keepOnlyInSource = isLocalLinkage(srcLinkage) ||
+                            isLinkOnceLinkage(srcLinkage) ||
+                            isAvailableExternallyLinkage(srcLinkage);
+
+    return pair.dst || shouldOverrideFromSrc() || !keepOnlyInSource;
+  }
+
+  FailureOr<bool> shouldLinkFromSource(ConflictPair pair) const {
+    auto srcLinkage = getLinkage(pair.src);
+    auto dstLinkage = getLinkage(pair.dst);
+
+    auto isDeclarationForLinker = [](Operation *op) {
+      if (isAvailableExternallyLinkage(getLinkage(op)))
+        return true;
+      return isDeclaration(op);
+    };
+
+    // Should we unconditionally use the src?
+    if (shouldOverrideFromSrc())
+      return true;
+
+    // We always have to add src if it has appending linkage.
+    if (isAppendingLinkage(srcLinkage) || isAppendingLinkage(dstLinkage))
+      return true;
+
+    if (isDeclarationForLinker(pair.src))
+      llvm_unreachable("Not implemented");
+
+    if (isDeclarationForLinker(pair.dst))
+      return true;
+
+    if (isCommonLinkage(srcLinkage)) {
+      if (isLinkOnceLinkage(dstLinkage) || isWeakLinkage(dstLinkage))
+        return true;
+      if (!isCommonLinkage(dstLinkage))
+        return true;
+    }
+
+    // TODO: This is not correct, should use some form of DataLayout concept
```

Comment:
For the LLVM dialect specifically, I guess this could be made correct without `DataLayout`?